robottelo/config/base.py

```diff
@@ -931,20 +931,42 @@ def __init__(self, *args, **kwargs):
         self.netmask = None
         self.gateway = None
         self.bridge = None
+        self.network = None
+        self.dhcp_ipam = None
+        self.dhcp_from = None
+        self.dhcp_to = None
+        self.dns_primary = None
 
     def read(self, reader):
         """Read Vlan Network settings."""
         self.subnet = reader.get('vlan_networking', 'subnet')
         self.netmask = reader.get('vlan_networking', 'netmask')
         self.gateway = reader.get('vlan_networking', 'gateway')
         self.bridge = reader.get('vlan_networking', 'bridge')
+        self.network = reader.get('vlan_networking', 'network')
+        self.dhcp_ipam = reader.get('vlan_networking', 'dhcp_ipam')
+        self.dhcp_from = reader.get('vlan_networking', 'dhcp_from')
+        self.dhcp_to = reader.get('vlan_networking', 'dhcp_to')
+        self.dns_primary = reader.get('vlan_networking', 'dns_primary')
 
     def validate(self):
         """Validate Vlan Network settings."""
         validation_errors = []
-        if not all(vars(self).values()):
+        if bool(self.bridge) == bool(self.network):
+            validation_errors.append(
+                'exactly one of the "bridge" or "network" parameters '
+                'must be specified')
+        if bool(self.dhcp_from) != bool(self.dhcp_to):
+            validation_errors.append(
+                'both or none of "dhcp_from", "dhcp_to" parameters '
+                'must be specified')
+            if self.dhcp_ipam and self.dhcp_ipam not in ['Internal DB', 'DHCP']:
+                validation_errors.append(
+                    '[vlan_networking] "dhcp_ipam" must be one of "Internal DB" or "DHCP"')
+        ignored = ['bridge', 'network', 'dhcp_ipam', 'dhcp_from', 'dhcp_to', 'dns_primary']
+        if not all(value for (key, value) in vars(self).items() if key not in ignored):
```

Comment:
Out of curiosity: how does it work?

I mean - `all` expects iterable. `value for (key, value) in vars(self).items() if key not in ignored`, in itself, is not valid Python - you need to surround it by `[]` or `()` to get list or generator comprehension. Here you omitted these and it still works. Why?

Reply:
Python will recognize it's a generator. Try it for yourself in a Python shell:
```python
all(value > 2 for value in [1, 2, 3])
all(value > 0 for value in [1, 2, 3])
```

I believe this is interpreted as `all((value > 0 for value in [1, 2, 3]))` and not `all([value > 0 for value in [1, 2, 3]])` but not entirely sure about that.

Reply:
@ekohl Yeah, I did notice it works, I just didn't know you can do something like that.

After searching a bit, it is added as part of [PEP 289](https://www.python.org/dev/peps/pep-0289/):

> if a function call has a single positional argument, it can be a generator expression without extra parentheses, but in all other cases you have to parenthesize it.

TIL